components/server/src/workspace/workspace-starter.ts

```diff
@@ -309,13 +311,20 @@ export class WorkspaceStarter {
                 ideConfig,
                 options.workspaceClass,
             );
+
             // we run the actual creation of a new instance in a distributed lock, to make sure we always only start one instance per workspace.
             await this.redisMutex.using(["workspace-start-" + workspace.id], 2000, async () => {
+                this.ensureShutdown(ctx, shutdownToken, workspace, undefined);
```

Comment:
I wonder if - instead of calling this method at every "checkpoint" - we could wait for a workspace-start to be finished. And if it's finished with an error, we land in the "failInstanceStart" logic below, which checks for the shutdownToken, or we succeed, and there is no need to check, but we can resolve the token. Does that make sense? :thinking:

Reply:
We have only 30 secs before SIGKILL there is no guarantee that it ever finishes?

Reply:
But we still have the same problem in the current state, no (if maybe reduced): What about individual sections (`await`) which take longer than 30s, like long image builds?

Reply:
`There is a hook installed below (shutdownToken.onCancellationRequested` which triggers immediately when ws/server are disposed. In other places I only check to fail earlier and prevent overriding with other messages.